File: WebApi/godotrl/src/main/java/com/example/godotrl/cst/perception/StateManager.java
package com.example.godotrl.cst.perception;

import br.unicamp.cst.core.entities.Codelet;
import br.unicamp.cst.core.entities.MemoryObject;
import com.example.godotrl.util.State;
import com.example.godotrl.util.Updater;
import com.example.godotrl.util.Vector2;

import java.util.ArrayList;

public class StateManager extends Codelet {
    MemoryObject positionMO = null;
    MemoryObject closestCarsMO = null;
    MemoryObject stateMO = null;
    MemoryObject lastStateMO = null;
    MemoryObject updateMO = null;

    @Override
    public void accessMemoryObjects() {
        positionMO = (MemoryObject) getInput("POSITION");
        closestCarsMO = (MemoryObject) getInput("CLOSEST_CARS");
        stateMO = (MemoryObject) getOutput("STATE");
        lastStateMO = (MemoryObject) getOutput("LAST_STATE");
        updateMO = (MemoryObject) getOutput("UPDATE");
    }

    @Override
    public void calculateActivation() {

    }

    @Override
    public void proc() {
        if (((Updater) updateMO.getI()).updateState()) {
            lastStateMO.setI(stateMO.getI());
            stateMO.setI(new State((Vector2) positionMO.getI(), (ArrayList<Vector2>) closestCarsMO.getI()));
        }
    }
}

File: GodotKotlin/MultiplatformFrogGame/src/main/kotlin/com/example/game/cst/perception/StateManager.kt
package com.example.game.cst.perception

import br.unicamp.cst.core.entities.Codelet
import br.unicamp.cst.core.entities.MemoryObject
import com.example.game.cst.State
import com.example.game.cst.behavior.Domain
import com.example.game.godot.Car
import godot.core.Vector2

class StateManager: Codelet() {
    var positionMO: MemoryObject? = null
    var closestCarsMO: MemoryObject? = null
    var stateMO: MemoryObject? = null
    var lastStateMO: MemoryObject? = null

    override fun accessMemoryObjects() {
        positionMO = getInput("POSITION") as MemoryObject
        closestCarsMO = getInput("CLOSEST_CARS") as MemoryObject
        stateMO = getOutput("STATE") as MemoryObject
        lastStateMO = getOutput("LAST_STATE") as MemoryObject
    }

    override fun calculateActivation() {
        TODO("Not yet implemented")
    }

    override fun proc() {
        var new_state = ArrayList<Domain<Double>>()
        var position = positionMO?.i as Vector2
        var closestCars = closestCarsMO?.i as ArrayList<Car>

        new_state.add(Domain<Double>(position.x))
        new_state.add(Domain<Double>(position.y))

        for (car: Car in closestCars) {
            new_state.add(Domain<Double>(car.cellPos.x))
            new_state.add(Domain<Double>(car.cellPos.y))
        }

        lastStateMO?.i = stateMO?.i
        stateMO?.i = new_state
    }
}